canonical/bundle-kubeflow
# fix/refactor bundle tests

# Summary
This PR fixes broken testing and refactors some tests for the bundle.

This change results in working CI that can be used on a locally deployed kubeflow (`tox -e tests -- --username XXX --password YYY`).  Note that this PR **does not** result result all problems in the github-driven integration tests as they have unrelated credential issues in AWS.

Other notes: I did not have a GPU available, so not sure if the `object_detection` pipeline currently works.  

Changes included:

### Changes to tests: 
* adds authentication flow (`tests/kf_authentication.py`) for pipeline tests (previously, auth was bypassed via port-forwarding in `run.sh`, but this stopped working)
* fixes selenium tests (mainly to adjust to minor webpage changes)
* fixes some pipeline tests
* improves the handling of dex credentials during testing, adding them as either command line arguments or pulled from environment variables during tests (previously some tests assumed default values or used hard coded ones)
* refactors `tests/test_kubectl.py` to check `juju status` of the bundle rather than look for specific pods or service accounts

### Changes to CI
* removes integration tests that were run on the github runner (these were run with the `kubeflow-edge` bundle, but the edge bundle is deprecated and no other bundle is small enough to run on github runners)
* pin AWS instance to a specific instance-type rather than resources to avoid over billing ([lp](https://bugs.launchpad.net/juju/+bug/1958574))
* set kubeflow/dex username/password as secrets to explicitly use them in the CI rather than rely on defaults

### Tests temporarily disabled
Some tests were temporarily disabled due to test failures.  These are documented in issues

* #431 `tests/test_selenium.py`/`test_notebook()` disabled due to upstream bug
* #432 `tests/test_pipeline.py`/`test_pipeline()`'s `mnist_pipeline` disabled as it was not working.  Not sure what is wrong
* #433 `tests/test_pipeline.py`/`test_pipeline()`'s `katib_pipeline` disabled as it was not working.  Not sure what is wrong

# Test instructions

(**see also the updated README.md file in the PR with more detailed instructions**)

* Deploy a local bundle-kubeflow (`juju deploy kubeflow --trust`)
* Patch istio (`kubectl patch role -n kubeflow istio-ingressgateway-operator -p '{"apiVersion":"rbac.authorization.k8s.io/v1","kind":"Role","metadata":{"name":"istio-ingressgateway-operator"},"rules":[{"apiGroups":["*"],"resources":["*"],"verbs":["*"]}]}'`)
* Set required config options
  ```
  juju config dex-auth static-username=user123@email.com
  juju config dex-auth static-password=user123
  juju config dex-auth public-url=http://10.64.140.43.nip.io
  juju config oidc-gatekeeper public-url=http://10.64.140.43.nip.io
  ```
* run tests (`tox -e tests -- -m lite --username user123@email.com --password user123`) (or `-m full`)
